module/VuFindSearch/src/VuFindSearch/Backend/Blender/Response/Json/RecordCollection.php

```diff
@@ -0,0 +1,441 @@
+        $this->mappings = $mappings;
+        $this->response = static::$template;
+        $this->initialResults = isset($this->config->Blending->initialResults)
+            ? $this->config->Blending->initialResults->toArray()
+            : [];
+    }
+
+    /**
+     * Initialize blended results
+     *
+     * Creates a record list from 0 to $limit
+     *
+     * @param array $collections Array of record collections
+     * @param int   $limit       Result limit
+     * @param int   $blockSize   Blending block size
+     * @param int   $totalCount  Total result count
+     *
+     * @return array Remaining records keyed by backend identifier
+     */
+    public function initBlended(
+        array $collections,
+        int $limit,
+        int $blockSize,
+        int $totalCount
+    ): array {
+        $this->response = static::$template;
+        $this->response['response']['numFound'] = $totalCount;
+        $this->rewind();
+
+        if (!$collections) {
+            return [];
+        }
+
+        $backendRecords = [];
+        foreach ($collections as $backendId => $collection) {
+            $records = $collection->getRecords();
+            $label = $this->config->Backends[$backendId];
+            foreach ($records as $record) {
+                $record->setSourceIdentifiers(
+                    $record->getSourceIdentifier(),
+                    $backendId
+                );
+                if ($label) {
+                    $record->addLabel($label, 'source');
+                }
+                $backendRecords[$backendId][] = $record;
+            }
+
+            foreach ($collection->getErrors() as $error) {
+                if (is_string($error) && $label) {
+                    $error = [
+                        'msg' => '%%error%% -- %%label%%',
+                        'tokens' => [
+                            '%%error%%' => $error,
+                            '%%label%%' => $label
+                        ],
+                        'translate' => true,
+                        'translateTokens' => true,
+                    ];
+                }
+                $this->addError($error);
+            }
+        }
+
+        $records = [];
+        $backendIds = array_keys($backendRecords);
+        // Filter out unavailable backends from initial results source list:
+        $initialResults = array_values(
+            array_filter(
+                $this->initialResults,
+                function ($s) use ($backendIds) {
+                    return in_array($s, $backendIds);
+                }
+            )
+        );
+        for ($pos = 0; $pos < $limit; $pos++) {
+            $backendId = $this->getBackendAtPosition(
+                $pos,
+                $blockSize,
+                $backendIds,
+                $initialResults
+            );
+            if (!empty($backendRecords[$backendId])) {
+                $this->add(array_shift($backendRecords[$backendId]), false);
```

Comment:
What happens when `$backendRecords[$backendId]` IS empty? It looks to me like this results in an incomplete page (since `$pos` keeps advancing regardless). Is that expected/normal, or am I missing something?

Reply:
This part just handles the initial results up to $limit with the defined positions. Then the backend fills the list up to the limit with the normal blocks, so there shouldn't be a case where the page is truncated.

Reply:
Maybe adding a comment at the top of this for loop to that effect will prevent me from forgetting and asking the question again in future. ;-)

Reply:
Added. :)